File: 17. Minimum Window Substring.py
# Topic: Minimum Window Substring (find smallest window containing all chars of pattern)

from collections import Counter

def min_window(s, t):
    if not t: return ""
    need = Counter(t)
    required = len(t)
    left = 0
    min_len = float('inf')
    start = 0
    for right, ch in enumerate(s):
        if need[ch] > 0:
            required -= 1
        need[ch] -= 1
        while required == 0:
            if right-left+1 < min_len:
                min_len = right-left+1
                start = left
            need[s[left]] += 1
            if need[s[left]] > 0:
                required += 1
            left += 1
    return "" if min_len == float('inf') else s[start:start+min_len]

if __name__ == '__main__':
    s = "ADOBECODEBANC"
    t = "ABC"
    print('MinWindow:', min_window(s,t))

# Time Complexity: O(n)
# Space Complexity: O(1) (alphabet-limited counter)

File: 17. Minimum Window Substring.cpp
// Topic: Minimum Window Substring (find smallest window containing all chars of pattern)
#include <bits/stdc++.h>
using namespace std;

string minWindow(string s, string t)
{
    if (t.empty())
        return "";
    vector<int> need(128, 0);
    for (char c : t) need[c]++;
    int required = t.size();
    int left = 0, minLen = INT_MAX, start = 0;
    for (int right = 0; right < s.size(); ++right)
    {
        if (need[s[right]] > 0)
            required--;
        need[s[right]]--;
        while (required == 0)
        {
            if (right - left + 1 < minLen)
            {
                minLen = right - left + 1;
                start = left;
            }
            need[s[left]]++;
            if (need[s[left]] > 0)
                required++;
            left++;
        }
    }
    return minLen == INT_MAX ? "" : s.substr(start, minLen);
}

int main()
{
    string s = "ADOBECODEBANC";
    string t = "ABC";
    cout << "MinWindow: " << minWindow(s, t) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) (constant-sized freq table)
